### PR TITLE
Rami4atm experiment

### PR DESCRIPTION
**Pull request by schunkes**
# Description

This PR adds the RAMI4ATM experiment and a specialized surface based on the `blendbsdf` mitsuba plugin.

Once the first round of feedback is done I will add more elaborate tests that compare the Rami4ATMExperiment's edge cases to the OneeDimExperiment and RamiExperiment classes.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

### CODE REVIEW

File: eradiate/experiments/_rami4atm.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                            toa = self.atmosphere.top
+                            target_point = [0.0, 0.0, toa.m] * toa.units
+                        else:
+                            target_point = [0.0, 0.0, 0.0] * ucc.get("length")
+
+    def _dataset_metadata(self, measure: Measure) -> t.Dict[str, str]:
+        result = super(Rami4ATMExperiment, self)._dataset_metadata(measure)
+
+        if isinstance(measure, DistantMeasure):
+            result["title"] = "Top-of-canopy simulation results"
```
Comment (leroyvn) [inline eradiate/experiments/_rami4atm.py:195]:
```suggestion
            result["title"] = "Top-of-atmosphere simulation results"
```

File: eradiate/experiments/_rami4atm.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        ),
+        doc=get_doc(Experiment, attrib="integrator", field="doc"),
+        type=get_doc(Experiment, attrib="integrator", field="type"),
+        init_type=get_doc(Experiment, attrib="integrator", field="init_type"),
+        default=":class:`VolPathIntegrator() <.VolPathIntegrator>`",
+    )
+
+    def __attrs_pre_init__(self):
+        # Only tested with monochromatic modes
+        supported_mode(ModeFlags.ANY_MONO)
```
Comment (leroyvn) [inline eradiate/experiments/_rami4atm.py:119]:
```suggestion
```

**Comment by schunkes** [comment]
Alright, here we go again. The docstring for the CentralPatchSurface might be a bit too short, but I wasn't sure how to describe it properly.

File: eradiate/contexts.py [hunk 3]
```diff
@@ -300,3 +300,18 @@ class KernelDictContext(Context):
...
+            default=None,
+            units=ucc.deferred("length"),
+            on_setattr=None,  # frozen classes can't use on_setattr
+        ),
+        doc="If relevant, value which must be used as the canopy width "
+        "(*e.g.* when the size of the central patch in a :class:`.CentralPatchSurface` "
+        "has to match a padded canopy).\n"
+        "\n"
+        "Unit-enabled field (default: ucc['length']).",
+        type="quantity, optional",
```
Comment (leroyvn) [inline eradiate/contexts.py:315]:
```suggestion
        type="quantity or None",
```

File: eradiate/experiments/_rami4atm.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    unit cell padding.
+
+    Notes
+    -----
+    A post-initialisation step will constrain the measure setup if a
+    :class:`.DistantMeasure` is used and no target is defined:
+
+    * if a canopy is defined, the target will be set to the canopy unit cell
+      (*i.e.* without its padding);
+    * if no canopy is defined, the target will be set to [0, 0, 0].
```
Comment (leroyvn) [inline eradiate/experiments/_rami4atm.py:37]: Doesn't this need an update?

File: eradiate/scenes/surface/_centralpatch.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                validators.auto_or(pinttr.validators.has_compatible_units),
+            ],
+            units=ucc.deferred("length"),
+        ),
+        doc="Central patch size. During kernel dictionary construction, ``AUTO`` "
+        "defaults defaults to one third of the overall surface size, unless a contextual "
+        "constraint (*e.g.* to match the size of an atmosphere or canopy) is applied.\n"
+        "\n"
+        "Unit-enabled field (default: ucc['length']).",
+        type="float or AUTO, optional",
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:77]:
```suggestion
        type="float or AUTO",
        init_type="float or AUTO, optional",
```

File: eradiate/scenes/surface/_centralpatch.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                if ctx.override_canopy_width
+                else self.patch_width
+            )
+            # the size of the central patch is one third of the overall size of the texture
+            return width / (3 * patch_width)
+
+    def bsdfs(self, ctx: KernelDictContext) -> KernelDict:
+        from mitsuba.core import ScalarTransform4f
+
+        returndict = {
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:100]: This should be a `KernelDict`.

File: eradiate/scenes/surface/_centralpatch.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+
+@surface_factory.register(type_id="central_patch")
+@parse_docs
+@attr.s
+class CentralPatchSurface(Surface):
+    """
+    Central patch surface element [``central_patch``]
+
+    This class creates a square surface to which two BSDFs will be attached.
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:25]: That's a bit too vague, would you mind elaborating a bit?

File: eradiate/scenes/surface/_centralpatch.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    The central patch will always be located in the center of the surface.
+    """
+
+    central_patch: Surface = documented(
+        attr.ib(
+            factory=LambertianSurface,
+            converter=surface_factory.convert,
+            validator=attr.validators.instance_of(Surface),
+        ),
+        doc="Inner surface specification. "
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:36]:
```suggestion
        doc="Central patch specification. "
```

File: eradiate/scenes/surface/_centralpatch.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        "This parameter can be specified as a dictionary which will be "
+        "interpreted by "
+        ":meth:`SurfaceFactory.convert() <.SurfaceFactory.convert>`.",
+        type=":class:`.Surface`",
+        init_type=":class:`.Surface` or dict",
+        default=":class:`LambertianSurface() <.LambertianSurface>`",
+    )
+
+    outer_surface: Surface = documented(
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:45]: How about `background_surface` instead, to clearly emphasise the idea of a background on which a patch is applied.

File: eradiate/scenes/surface/_centralpatch.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        doc="Outer surface specification. "
+        "This parameter can be specified as a dictionary which will be "
+        "interpreted by "
+        ":meth:`SurfaceFactory.convert() <.SurfaceFactory.convert>`.",
+        type=":class:`.Surface`",
+        init_type=":class:`.Surface` or dict",
+        default=":class:`LambertianSurface() <.LambertianSurface>`",
+    )
+
+    patch_width: t.Union[pint.Quantity, AutoType] = documented(
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:60]: The more I think about it, the more I believe that this parameter can be removed. You can just use `central_patch.width` instead. What do you think?

**Review by leroyvn — CHANGES_REQUESTED** [review]
Hi @schunkes, we're almost there. Just a couple of small changes and we'll be good to go!

File: eradiate/scenes/surface/_centralpatch.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            ],
+            units=ucc.deferred("length"),
+        ),
+        doc="Central patch size. During kernel dictionary construction, ``AUTO`` "
+        "defaults defaults to one third of the overall surface size, unless a contextual "
+        "constraint (*e.g.* to match the size of an atmosphere or canopy) is applied.\n"
+        "\n"
+        "Unit-enabled field (default: ucc['length']).",
+        type="float or AUTO",
+        default="float or AUTO, optional",
```
Comment (leroyvn) [inline eradiate/scenes/surface/_centralpatch.py:83]:
```suggestion
        init_type="float or AUTO, optional",
        default="AUTO",
```

**Comment by schunkes** [comment]
I implemented the things we discussed yesterday. 

Additionally I added a separate commit, where I set the default ray origin to cover the entire horizontal extent of the scene and be located a small distance above either the surface, canopy or atmosphere, whichever is highest.
For that I had to allow the ray origin to be set to `TargetOriginRectangle` and I don't know if there was a reason for not allowing that, which is why I kept this as a separate commit and we can leave it out if you think it should not be used.
